packages/rest/src/__tests__/integration/request-context.integration.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {ApplicationConfig} from '@loopback/core';
+import {
+  Client,
+  createRestAppClient,
+  expect,
+  givenHttpServerConfig,
+  supertest,
+  httpsGetAsync,
+} from '@loopback/testlab';
+import * as express from 'express';
+import {RequestContext} from '../../request-context';
+import {RestApplication} from '../../rest.application';
+import {RestServerConfig} from '../../rest.server';
+import {DefaultSequence} from '../../sequence';
+
+let app: RestApplication;
+let client: Client;
+let observedCtx: RequestContext;
+
+describe('RequestContext', () => {
+  beforeEach(setup);
+  afterEach(teardown);
+
+  describe('requestedProtocol', () => {
+    it('defaults to "http"', async () => {
+      await givenRunningAppWithClient();
+      await client.get('/products').expect(200);
+      expect(observedCtx.requestedProtocol).to.equal('http');
+    });
+
+    it('honors "x-forwarded-proto" header', async () => {
+      await givenRunningAppWithClient();
+      await client
+        .get('/products')
+        .set('x-forwarded-proto', 'https')
+        .expect(200);
+      expect(observedCtx.requestedProtocol).to.equal('https');
+    });
+
+    it('honors protocol provided by Express request', async () => {
+      await givenRunningAppWithClient({protocol: 'https'});
+      expect(app.restServer.url).to.startWith('https:');
+      // supertest@3 fails with Error: self signed certificate
+      // FIXME(bajtos) rework this code once we upgrade to supertest@4
+      // await client.get('/products').trustLocalhost().expect(200);
+      await httpsGetAsync(app.restServer.url + '/products');
+      expect(observedCtx.requestedProtocol).to.equal('https');
+    });
+  });
+
+  describe('basePath', () => {
+    it('defaults to an empty string', async () => {
+      await givenRunningAppWithClient();
+      await client.get('/products').expect(200);
+      expect(observedCtx.basePath).to.equal('');
+    });
+
+    it('honors baseUrl when mounted on a sub-path', async () => {
+      const lbApp = new RestApplication();
+      lbApp.handler(contextObservingHandler);
+
+      const expressApp = express();
+      expressApp.use('/api', lbApp.requestHandler);
```

Comment:
What's going to happen if `lbApp.basePath('/v1')` is called?

Reply:
We probably need to have a test to cover the case above.

Reply:
I agree it's a use case we should cover by our test suite and possibly fix.

At the same time, I feel it's out of scope of this pull request. My intention is to move certain computed properties from RestServer to RequestContext while preserving their behavior.

I'll open a new PR to take a look at this.

Reply:
See https://github.com/strongloop/loopback-next/pull/2589